examples/xgboost-comprehensive/sim.py

```diff
@@ -0,0 +1,156 @@
+import warnings
+from logging import INFO
+
+import flwr as fl
+from flwr_datasets import FederatedDataset
+from flwr.common.logger import log
+from flwr.server.strategy import FedXgbBagging, FedXgbCyclic
+
+from dataset import (
+    instantiate_partitioner,
+    train_test_split,
+    transform_dataset_to_dmatrix,
+    resplit,
+)
+from utils import (
+    sim_args_parser,
+    NUM_LOCAL_ROUND,
+    BST_PARAMS,
+)
+from server_utils import (
+    eval_config,
+    fit_config,
+    evaluate_metrics_aggregation,
+    get_evaluate_fn,
+    CyclicClientManager,
+)
+from client_utils import XgbClient
+
+
+warnings.filterwarnings("ignore", category=UserWarning)
+
+
+def get_client_fn(fds, args, params, num_local_round):
+    """Return a function to construct a client.
+
+    The VirtualClientEngine will exectue this function whenever a client is sampled by
+    the strategy to participate.
+    """
+
+    def client_fn(cid: str) -> fl.client.Client:
+        """Construct a FlowerClient with its own dataset partition."""
+
+        # Extract partition for client with node_id = cid
+        partition = fds.load_partition(node_id=int(cid), split="train")
+        partition.set_format("numpy")
+
+        if args.centralised_eval_client:
+            # Use centralised test set for evaluation
+            train_data = partition
+            valid_data = fds.load_full("test")
+            valid_data.set_format("numpy")
+            num_train = train_data.shape[0]
+            num_val = valid_data.shape[0]
+        else:
+            # Train/test splitting
+            train_data, valid_data, num_train, num_val = train_test_split(
+                partition, test_fraction=args.test_fraction, seed=args.seed
+            )
+
+        # Reformat data to DMatrix for xgboost
+        train_dmatrix = transform_dataset_to_dmatrix(train_data)
+        valid_dmatrix = transform_dataset_to_dmatrix(valid_data)
```

Comment:
@yan-gao-GY , can we do this offline? (i.e. before calling `start_simulation`). I see this takes ~20s for each partition, slowing down the simulation by a lot.

Reply:
if we want to do the data re-formatting before partitioning, we need to first load the full dataset. I'm not sure how to do partitioning after loading the full dataset. I think it changes the logic of Flower Datasets. So I think we need to do something insides of Flower Datasets to achieve this. What do you think? Any better way? @jafermarq